File: ts/api/src/config/configLoader.ts
/**
 * YAML Configuration Loader (TypeScript)
 * Centralized config loading for networks, tokens, pricing, deployment
 */
import { readFileSync } from "fs";
import { resolve } from "path";
import yaml from "yaml";

// Config directory path (relative to repo root)
const CONFIG_DIR = resolve(__dirname, "../../../../config");

export interface NetworkConfig {
  chain_id: number;
  rpc_urls: string[];
  explorer: string;
  currency: string;
  features: {
    batch_deployment: boolean;
    eigenda: boolean;
  };
}

export interface TokenAddresses {
  [network: string]: string;
}

export interface TokenMetadata {
  name: string;
  symbol: string;
  decimals: number;
  type: string;
}

export interface PricingConfig {
  contract_types: Record<string, number>;
  workflow_tiers: Record<string, number>;
  tasks: Record<string, number>;
  network_multipliers: Record<string, number>;
  model_multipliers: Record<string, number>;
  complexity_multipliers: Record<string, number>;
}

export interface DeploymentConfig {
  strategies: {
    erc4337: {
      enabled: boolean;
      networks: string[];
      entrypoint_address: string;
      paymaster_required: boolean;
    };
    server_wallet: {
      enabled: boolean;
      networks: string[];
      max_gas_price_gwei: number;
    };
  };
  gas_settings: Record<string, {
    max_priority_fee_gwei: number;
    max_fee_gwei: number;
    gas_limit_multiplier: number;
  }>;
  verification: {
    auto_verify: boolean;
    explorers: Record<string, string>;
  };
  wallets?: {
    merchant?: {
      address: string;
      description: string;
      networks: string[];
    };
    server_deployer?: {
      address: string;
      description: string;
      networks: string[];
    };
    erc4337?: {
      entrypoint: string;
      factory: string;
      description: string;
    };
  };
}

export interface LLMProviderConfig {
  enabled: boolean;
  priority: number;
  models: {
    default: string;
  };
  generation: {
    temperature: number;
    max_output_tokens?: number;
    max_tokens?: number;
    thinking_budget?: number | null;
    timeout_seconds: number;
  };
  embedding?: {
    model: string;
    dimensions: number;
    timeout_seconds: number;
  };
  rate_limits: {
    requests_per_minute: number;
    tokens_per_minute: number;
  };
  features: {
    multimodal: boolean;
    function_calling: boolean;
    streaming: boolean;
    json_mode: boolean;
  };
}

export interface LLMRoutingConfig {
  strategy: string;
  tasks: Record<string, {
    primary: string;
    fallback: string[];
    timeout: number;
    retry_attempts: number;
  }>;
  fallback: {
    enabled: boolean;
    max_retries: number;
    backoff_multiplier: number;
    retry_on_errors: string[];
  };
}

export interface LLMConfig {
  providers: Record<string, LLMProviderConfig>;
  routing: LLMRoutingConfig;
  pricing: Record<string, {
    input: number;
    output: number;
  }>;
  features: {
    enable_acontext: boolean;
    enable_multi_model_router: boolean;
    enable_cost_tracking: boolean;
    enable_latency_monitoring: boolean;
    enable_automatic_fallback: boolean;
    prefer_streaming: boolean;
  };
  global: {
    default_provider: string;
    embedding_provider: string;
    max_concurrent_requests: number;
    request_queue_size: number;
    cache_responses: boolean;
    cache_ttl_seconds: number;
  };
}

export interface X402Config {
  enabled: boolean;
  version: string;
  protocol_standard: string;
  supported_networks: string[];
  network_settings: Record<string, {
    enabled: boolean;
    testnet: boolean;
    confirmation_blocks: number;
    timeout_seconds: number;
  }>;
  verifier: {
    service_url_env: string;
    default_url: string;
    timeout_seconds: number;
    retry_attempts: number;
    retry_delay_seconds: number;
  };
  settlement: {
    auto_settle: boolean;
    batch_settlements: boolean;
    batch_size: number;
    batch_timeout_minutes: number;
    min_amount_usdc: number;
    max_amount_usdc: number;
    use_gas_optimization: boolean;
    max_gas_price_gwei: number;
  };
  payment_tokens: Record<string, {
    enabled: boolean;
    symbol: string;
    decimals: number;
  }>;
  pricing: {
    contract_generation: Record<string, number>;
    workflow: Record<string, number>;
    deployment: Record<string, number>;
    tasks: Record<string, number>;
  };
  payment_flow: {
    verify_before_execution: boolean;
    require_confirmation: boolean;
    confirmation_blocks: number;
    methods: string[];
    default_method: string;
  };
  refunds: {
    enabled: boolean;
    auto_refund_on_failure: boolean;
    refund_timeout_hours: number;
    conditions: string[];
    manual_review_amount_usdc: number;
  };
  fraud_prevention: {
    enabled: boolean;
    rate_limits: {
      requests_per_hour: number;
      requests_per_day: number;
      max_amount_per_day_usdc: number;
    };
    use_blacklist: boolean;
    use_whitelist: boolean;
    require_wallet_age_days: number;
    require_min_balance_usdc: number;
  };
}

class ConfigLoader {
  private cache: Map<string, any> = new Map();

  private loadYaml<T>(filename: string): T {
    if (this.cache.has(filename)) {
      return this.cache.get(filename) as T;
    }

    const filepath = resolve(CONFIG_DIR, filename);
    try {
      const content = readFileSync(filepath, "utf-8");
      const parsed = yaml.parse(content);
      this.cache.set(filename, parsed);
      return parsed as T;
    } catch (error) {
      throw new Error(`Failed to load config file ${filename}: ${error}`);
    }
  }

  getNetworks(): Record<string, NetworkConfig> {
    const data = this.loadYaml<{ networks: Record<string, NetworkConfig> }>("networks.yaml");
    return data.networks || {};
  }

  getNetwork(networkId: string): NetworkConfig | undefined {
    const networks = this.getNetworks();
    return networks[networkId];
  }

  getTokens(): Record<string, TokenAddresses> {
    const data = this.loadYaml<{ tokens: Record<string, TokenAddresses> }>("tokens.yaml");
    return data.tokens || {};
  }

  getTokenAddress(tokenSymbol: string, networkId: string): string | undefined {
    const tokens = this.getTokens();
    const tokenData = tokens[tokenSymbol.toLowerCase()];
    return tokenData?.[networkId];
  }

  getTokenMetadata(tokenSymbol: string): TokenMetadata | undefined {
    const data = this.loadYaml<{ metadata: Record<string, TokenMetadata> }>("tokens.yaml");
    return data.metadata?.[tokenSymbol.toLowerCase()];
  }

  getPricing(): PricingConfig {
    return this.loadYaml<PricingConfig>("pricing.yaml");
  }

  getContractPrice(contractType: string): number {
    const pricing = this.getPricing();
    return pricing.contract_types?.[contractType] ?? 0.15;
  }

  getWorkflowPrice(tier: string): number {
    const pricing = this.getPricing();
    return pricing.workflow_tiers?.[tier] ?? 0.02;
  }

  getNetworkMultiplier(networkId: string): number {
    const pricing = this.getPricing();
    return pricing.network_multipliers?.[networkId] ?? 1.0;
  }

  getModelMultiplier(modelName: string): number {
    const pricing = this.getPricing();
    return pricing.model_multipliers?.[modelName] ?? 1.0;
  }

  getDeploymentConfig(): DeploymentConfig {
    return this.loadYaml<DeploymentConfig>("deployment.yaml");
  }

  getDeploymentStrategy(networkId: string): "erc4337" | "server_wallet" | "thirdweb_engine" {
    const config = this.getDeploymentConfig();

    if (config.strategies.erc4337.networks.includes(networkId)) {
      return "erc4337";
    }

    if (config.strategies.server_wallet.networks.includes(networkId)) {
      return "server_wallet";
    }

    return "server_wallet"; // Default
  }

  getGasSettings(networkId: string) {
    const config = this.getDeploymentConfig();
    return config.gas_settings?.[networkId];
  }

  getExplorerApi(networkId: string): string | undefined {
    const config = this.getDeploymentConfig();
    return config.verification.explorers?.[networkId];
  }

  listSupportedNetworks(): string[] {
    return Object.keys(this.getNetworks());
  }

  listSupportedTokens(): string[] {
    return Object.keys(this.getTokens());
  }

  isTestnet(networkId: string): boolean {
    return (
      networkId.includes("_testnet") ||
      networkId.includes("sepolia") ||
      networkId.includes("fuji") ||
      networkId.includes("calibration") ||
      networkId.includes("amoy")
    );
  }

  // ========================================================================
  // LLM Configuration Methods
  // ========================================================================

  getLLMConfig(): LLMConfig {
    return this.loadYaml<LLMConfig>("llm.yaml");
  }

  getLLMProviderConfig(providerName: string): LLMProviderConfig | undefined {
    const config = this.getLLMConfig();
    return config.providers?.[providerName];
  }

  getEnabledLLMProviders(): string[] {
    const config = this.getLLMConfig();
    const providers = config.providers || {};

    const enabled = Object.entries(providers)
      .filter(([_, data]) => data.enabled)
      .map(([name, data]) => ({ name, priority: data.priority || 99 }))
      .sort((a, b) => a.priority - b.priority);

    return enabled.map((p) => p.name);
  }

  getPrimaryLLMProvider(): string {
    const enabled = this.getEnabledLLMProviders();
    return enabled[0] || "gemini"; // Default fallback
  }

  getLLMModel(providerName: string): string {
    const provider = this.getLLMProviderConfig(providerName);
    return provider?.models?.default || "";
  }

  getLLMGenerationConfig(providerName: string): LLMProviderConfig["generation"] | undefined {
    const provider = this.getLLMProviderConfig(providerName);
    return provider?.generation;
  }

  getLLMRoutingConfig(): LLMRoutingConfig {
    const config = this.getLLMConfig();
    return config.routing;
  }

  getLLMRoutingForTask(taskName: string) {
    const routing = this.getLLMRoutingConfig();
    return routing.tasks?.[taskName];
  }

  getLLMPricing(providerName: string) {
    const config = this.getLLMConfig();
    return config.pricing?.[providerName];
  }

  // ========================================================================
  // Wallet Configuration Methods
  // ========================================================================

  getMerchantWalletAddress(): string {
    const config = this.getDeploymentConfig();
    return config.wallets?.merchant?.address || "";
  }

  getServerDeployerAddress(): string {
    const config = this.getDeploymentConfig();
    return config.wallets?.server_deployer?.address || "";
  }

  getERC4337Entrypoint(): string {
    const config = this.getDeploymentConfig();
    return config.wallets?.erc4337?.entrypoint || "0x0000000071727De22E5E9d8BAf0edAc6f37da032";
  }

  getERC4337Factory(): string {
    const config = this.getDeploymentConfig();
    return config.wallets?.erc4337?.factory || "";
  }

  // ========================================================================
  // x402 Configuration Methods
  // ========================================================================

  getX402Config(): X402Config {
    return this.loadYaml<X402Config>("x402.yaml");
  }

  isX402Enabled(): boolean {
    const config = this.getX402Config();
    return config.enabled ?? false;
  }

  getX402SupportedNetworks(): string[] {
    const config = this.getX402Config();
    return config.supported_networks || [];
  }

  isNetworkX402Supported(networkId: string): boolean {
    return this.getX402SupportedNetworks().includes(networkId);
  }

  getX402NetworkSettings(networkId: string) {
    const config = this.getX402Config();
    return config.network_settings?.[networkId];
  }

  getX402VerifierConfig() {
    const config = this.getX402Config();
    return config.verifier;
  }

  getX402Pricing(category: string, tier: string = "basic"): number {
    const config = this.getX402Config();
    const pricing = config.pricing?.[category as keyof typeof config.pricing];
    if (pricing && typeof pricing === "object") {
      return (pricing as Record<string, number>)[tier] ?? 0.0;
    }
    return 0.0;
  }

  getX402PaymentMethods(): string[] {
    const config = this.getX402Config();
    return config.payment_flow?.methods || ["erc4337_bundler"];
  }

  getX402DefaultPaymentMethod(): string {
    const config = this.getX402Config();
    return config.payment_flow?.default_method || "erc4337_bundler";
  }

  isX402RefundEnabled(): boolean {
    const config = this.getX402Config();
    return config.refunds?.enabled ?? true;
  }

  getX402RateLimits() {
    const config = this.getX402Config();
    return config.fraud_prevention?.rate_limits;
  }

  clearCache(): void {
    this.cache.clear();
  }
}

// Singleton instance
let _configLoader: ConfigLoader | null = null;

export function getConfigLoader(): ConfigLoader {
  if (!_configLoader) {
    _configLoader = new ConfigLoader();
  }
  return _configLoader;
}

// Convenience exports
export const getNetwork = (networkId: string) => getConfigLoader().getNetwork(networkId);
export const getTokenAddress = (tokenSymbol: string, networkId: string) =>
  getConfigLoader().getTokenAddress(tokenSymbol, networkId);
export const getContractPrice = (contractType: string) => getConfigLoader().getContractPrice(contractType);
export const getDeploymentStrategy = (networkId: string) => getConfigLoader().getDeploymentStrategy(networkId);
export const listNetworks = () => getConfigLoader().listSupportedNetworks();
export const isTestnet = (networkId: string) => getConfigLoader().isTestnet(networkId);

// LLM convenience exports
export const getLLMProviderConfig = (providerName: string) => getConfigLoader().getLLMProviderConfig(providerName);
export const getPrimaryLLMProvider = () => getConfigLoader().getPrimaryLLMProvider();
export const getLLMModel = (providerName: string) => getConfigLoader().getLLMModel(providerName);

// Wallet convenience exports
export const getMerchantWalletAddress = () => getConfigLoader().getMerchantWalletAddress();
export const getERC4337Entrypoint = () => getConfigLoader().getERC4337Entrypoint();

// x402 convenience exports
export const isX402Enabled = () => getConfigLoader().isX402Enabled();
export const getX402SupportedNetworks = () => getConfigLoader().getX402SupportedNetworks();
export const isNetworkX402Supported = (networkId: string) => getConfigLoader().isNetworkX402Supported(networkId);
export const getX402Pricing = (category: string, tier?: string) => getConfigLoader().getX402Pricing(category, tier);

File: hyperagent/core/config_loader.py
"""
YAML Configuration Loader
Centralized config loading for networks, tokens, pricing, deployment
"""
import os
from pathlib import Path
from typing import Any, Dict, Optional
import yaml
from functools import lru_cache

# Config directory path (relative to repo root)
CONFIG_DIR = Path(__file__).parent.parent.parent / "config"


class ConfigLoader:
    """Load and cache YAML configuration files"""
    
    def __init__(self, config_dir: Optional[Path] = None):
        self.config_dir = config_dir or CONFIG_DIR
        self._cache: Dict[str, Any] = {}
    
    @lru_cache(maxsize=10)
    def load_yaml(self, filename: str) -> Dict[str, Any]:
        """Load YAML file with caching"""
        filepath = self.config_dir / filename
        
        if not filepath.exists():
            raise FileNotFoundError(f"Config file not found: {filepath}")
        
        with open(filepath, "r", encoding="utf-8") as f:
            return yaml.safe_load(f) or {}
    
    def get_networks(self) -> Dict[str, Any]:
        """Load networks.yaml"""
        return self.load_yaml("networks.yaml").get("networks", {})
    
    def get_network(self, network_id: str) -> Optional[Dict[str, Any]]:
        """Get specific network configuration"""
        networks = self.get_networks()
        return networks.get(network_id)
    
    def get_tokens(self) -> Dict[str, Any]:
        """Load tokens.yaml"""
        return self.load_yaml("tokens.yaml").get("tokens", {})
    
    def get_token_address(self, token_symbol: str, network_id: str) -> Optional[str]:
        """Get token contract address for specific network"""
        tokens = self.get_tokens()
        token_data = tokens.get(token_symbol.lower(), {})
        return token_data.get(network_id)
    
    def get_token_metadata(self, token_symbol: str) -> Dict[str, Any]:
        """Get token metadata (name, symbol, decimals)"""
        metadata = self.load_yaml("tokens.yaml").get("metadata", {})
        return metadata.get(token_symbol.lower(), {})
    
    def get_pricing(self) -> Dict[str, Any]:
        """Load pricing.yaml"""
        return self.load_yaml("pricing.yaml")
    
    def get_contract_price(self, contract_type: str) -> float:
        """Get price for contract type"""
        pricing = self.get_pricing()
        return pricing.get("contract_types", {}).get(contract_type, 0.15)
    
    def get_workflow_price(self, tier: str) -> float:
        """Get price for workflow tier"""
        pricing = self.get_pricing()
        return pricing.get("workflow_tiers", {}).get(tier, 0.02)
    
    def get_network_multiplier(self, network_id: str) -> float:
        """Get pricing multiplier for network"""
        pricing = self.get_pricing()
        return pricing.get("network_multipliers", {}).get(network_id, 1.0)
    
    def get_model_multiplier(self, model_name: str) -> float:
        """Get pricing multiplier for LLM model"""
        pricing = self.get_pricing()
        multipliers = pricing.get("model_multipliers", {})
        return multipliers.get(model_name, 1.0)
    
    def get_deployment_config(self) -> Dict[str, Any]:
        """Load deployment.yaml"""
        return self.load_yaml("deployment.yaml")
    
    def get_deployment_strategy(self, network_id: str) -> str:
        """Determine deployment strategy for network"""
        config = self.get_deployment_config()
        strategies = config.get("strategies", {})
        
        # Check ERC-4337 support
        if network_id in strategies.get("erc4337", {}).get("networks", []):
            return "erc4337"
        
        # Check server wallet support
        if network_id in strategies.get("server_wallet", {}).get("networks", []):
            return "server_wallet"
        
        # Default to server wallet
        return "server_wallet"
    
    def get_gas_settings(self, network_id: str) -> Dict[str, Any]:
        """Get gas configuration for network"""
        config = self.get_deployment_config()
        gas_settings = config.get("gas_settings", {})
        return gas_settings.get(network_id, {})
    
    def get_explorer_api(self, network_id: str) -> Optional[str]:
        """Get block explorer API URL for contract verification"""
        config = self.get_deployment_config()
        explorers = config.get("verification", {}).get("explorers", {})
        return explorers.get(network_id)
    
    def list_supported_networks(self) -> list[str]:
        """List all configured networks"""
        return list(self.get_networks().keys())
    
    def list_supported_tokens(self) -> list[str]:
        """List all configured tokens"""
        return list(self.get_tokens().keys())
    
    def is_testnet(self, network_id: str) -> bool:
        """Check if network is a testnet"""
        return "_testnet" in network_id or "sepolia" in network_id or "fuji" in network_id or "calibration" in network_id or "amoy" in network_id
    
    # ========================================================================
    # LLM Configuration Methods
    # ========================================================================
    
    def get_llm_config(self) -> Dict[str, Any]:
        """Load llm.yaml"""
        return self.load_yaml("llm.yaml")
    
    def get_llm_provider_config(self, provider_name: str) -> Optional[Dict[str, Any]]:
        """Get configuration for specific LLM provider (gemini, openai, anthropic, together)"""
        config = self.get_llm_config()
        providers = config.get("providers", {})
        return providers.get(provider_name)
    
    def get_enabled_llm_providers(self) -> list[str]:
        """Get list of enabled LLM providers in priority order"""
        config = self.get_llm_config()
        providers = config.get("providers", {})
        
        enabled = [
            (name, data.get("priority", 99))
            for name, data in providers.items()
            if data.get("enabled", False)
        ]
        
        enabled.sort(key=lambda x: x[1])
        return [name for name, _ in enabled]
    
    def get_primary_llm_provider(self) -> str:
        """Get primary (highest priority) LLM provider"""
        enabled = self.get_enabled_llm_providers()
        if enabled:
            return enabled[0]
        return "gemini"  # Default fallback
    
    def get_llm_model(self, provider_name: str) -> str:
        """Get default model for provider"""
        provider_config = self.get_llm_provider_config(provider_name)
        if provider_config:
            return provider_config.get("models", {}).get("default", "")
        return ""
    
    def get_llm_generation_config(self, provider_name: str) -> Dict[str, Any]:
        """Get generation settings for provider (temperature, max_tokens, etc.)"""
        provider_config = self.get_llm_provider_config(provider_name)
        if provider_config:
            return provider_config.get("generation", {})
        return {}
    
    def get_llm_routing_config(self) -> Dict[str, Any]:
        """Get multi-model routing configuration"""
        config = self.get_llm_config()
        return config.get("routing", {})
    
    def get_llm_routing_for_task(self, task_name: str) -> Dict[str, Any]:
        """Get routing configuration for specific task (solidity_codegen, gas_optimization, etc.)"""
        routing = self.get_llm_routing_config()
        tasks = routing.get("tasks", {})
        return tasks.get(task_name, {})
    
    def get_llm_pricing(self, provider_name: str) -> Dict[str, float]:
        """Get pricing for LLM provider (input/output per 1M tokens)"""
        config = self.get_llm_config()
        pricing = config.get("pricing", {})
        return pricing.get(provider_name, {})
    
    # ========================================================================
    # Wallet Configuration Methods
    # ========================================================================
    
    def get_merchant_wallet_address(self) -> str:
        """Get merchant wallet address for x402 payments"""
        config = self.get_deployment_config()
        wallets = config.get("wallets", {})
        merchant = wallets.get("merchant", {})
        return merchant.get("address", "")
    
    def get_server_deployer_address(self) -> str:
        """Get server wallet address for deployments (public address only)"""
        config = self.get_deployment_config()
        wallets = config.get("wallets", {})
        server = wallets.get("server_deployer", {})
        return server.get("address", "")
    
    def get_erc4337_entrypoint(self) -> str:
        """Get ERC-4337 EntryPoint address"""
        config = self.get_deployment_config()
        wallets = config.get("wallets", {})
        erc4337 = wallets.get("erc4337", {})
        return erc4337.get("entrypoint", "0x0000000071727De22E5E9d8BAf0edAc6f37da032")
    
    def get_erc4337_factory(self) -> str:
        """Get ERC-4337 Account Factory address"""
        config = self.get_deployment_config()
        wallets = config.get("wallets", {})
        erc4337 = wallets.get("erc4337", {})
        return erc4337.get("factory", "")
    
    # ========================================================================
    # x402 Configuration Methods
    # ========================================================================
    
    def get_x402_config(self) -> Dict[str, Any]:
        """Load x402.yaml"""
        return self.load_yaml("x402.yaml")
    
    def is_x402_enabled(self) -> bool:
        """Check if x402 payments are enabled globally"""
        config = self.get_x402_config()
        return config.get("enabled", False)
    
    def get_x402_supported_networks(self) -> list[str]:
        """Get list of networks supporting x402 payments"""
        config = self.get_x402_config()
        return config.get("supported_networks", [])
    
    def is_network_x402_supported(self, network_id: str) -> bool:
        """Check if network supports x402 payments"""
        return network_id in self.get_x402_supported_networks()
    
    def get_x402_network_settings(self, network_id: str) -> Dict[str, Any]:
        """Get x402 settings for specific network"""
        config = self.get_x402_config()
        network_settings = config.get("network_settings", {})
        return network_settings.get(network_id, {})
    
    def get_x402_verifier_config(self) -> Dict[str, Any]:
        """Get x402 payment verifier configuration"""
        config = self.get_x402_config()
        return config.get("verifier", {})
    
    def get_x402_pricing(self, category: str, tier: str = "basic") -> float:
        """Get x402 price for category and tier"""
        config = self.get_x402_config()
        pricing = config.get("pricing", {})
        category_pricing = pricing.get(category, {})
        return category_pricing.get(tier, 0.0)
    
    def get_x402_payment_methods(self) -> list[str]:
        """Get list of supported x402 payment methods"""
        config = self.get_x402_config()
        payment_flow = config.get("payment_flow", {})
        return payment_flow.get("methods", ["erc4337_bundler"])
    
    def get_x402_default_payment_method(self) -> str:
        """Get default x402 payment method"""
        config = self.get_x402_config()
        payment_flow = config.get("payment_flow", {})
        return payment_flow.get("default_method", "erc4337_bundler")
    
    def is_x402_refund_enabled(self) -> bool:
        """Check if x402 refunds are enabled"""
        config = self.get_x402_config()
        refunds = config.get("refunds", {})
        return refunds.get("enabled", True)
    
    def get_x402_rate_limits(self) -> Dict[str, Any]:
        """Get x402 rate limits per wallet"""
        config = self.get_x402_config()
        fraud = config.get("fraud_prevention", {})
        return fraud.get("rate_limits", {})
    
    def get_x402_protocol_standard(self) -> str:
        """Get x402 protocol standard (EIP-7702 or ERC-4337)"""
        config = self.get_x402_config()
        return config.get("protocol_standard", "EIP-7702")
    
    def get_x402_gas_sponsorship_config(self) -> Dict[str, Any]:
        """Get Thirdweb gas sponsorship configuration"""
        config = self.get_x402_config()
        integrations = config.get("integrations", {})
        thirdweb = integrations.get("thirdweb", {})
        return thirdweb.get("gas_sponsorship", {})


# Singleton instance
_config_loader: Optional[ConfigLoader] = None


def get_config_loader() -> ConfigLoader:
    """Get or create singleton ConfigLoader instance"""
    global _config_loader
    if _config_loader is None:
        _config_loader = ConfigLoader()
    return _config_loader


# Convenience functions
def get_network(network_id: str) -> Optional[Dict[str, Any]]:
    """Get network configuration"""
    return get_config_loader().get_network(network_id)


def get_token_address(token_symbol: str, network_id: str) -> Optional[str]:
    """Get token address for network"""
    return get_config_loader().get_token_address(token_symbol, network_id)


def get_contract_price(contract_type: str) -> float:
    """Get contract generation price"""
    return get_config_loader().get_contract_price(contract_type)


def get_deployment_strategy(network_id: str) -> str:
    """Get deployment strategy for network"""
    return get_config_loader().get_deployment_strategy(network_id)


def list_networks() -> list[str]:
    """List all supported networks"""
    return get_config_loader().list_supported_networks()


def is_testnet(network_id: str) -> bool:
    """Check if network is testnet"""
    return get_config_loader().is_testnet(network_id)


def get_llm_provider_config(provider_name: str) -> Optional[Dict[str, Any]]:
    """Get LLM provider configuration"""
    return get_config_loader().get_llm_provider_config(provider_name)


def get_primary_llm_provider() -> str:
    """Get primary LLM provider"""
    return get_config_loader().get_primary_llm_provider()


def get_merchant_wallet_address() -> str:
    """Get merchant wallet address for x402 payments"""
    return get_config_loader().get_merchant_wallet_address()


def is_x402_enabled() -> bool:
    """Check if x402 payments are enabled"""
    return get_config_loader().is_x402_enabled()


def get_x402_supported_networks() -> list[str]:
    """Get x402 supported networks"""
    return get_config_loader().get_x402_supported_networks()


def is_network_x402_supported(network_id: str) -> bool:
    """Check if network supports x402"""
    return get_config_loader().is_network_x402_supported(network_id)


def get_x402_pricing(category: str, tier: str = "basic") -> float:
    """Get x402 pricing for category and tier"""
    return get_config_loader().get_x402_pricing(category, tier)


def get_x402_config() -> Dict[str, Any]:
    """Get x402 configuration"""
    return get_config_loader().get_x402_config()


def get_x402_protocol_standard() -> str:
    """Get x402 protocol standard"""
    return get_config_loader().get_x402_protocol_standard()

